File: native/ios/GlyphisShell/ImageLoader.swift
import Foundation
import ImageIO

/// Loads images from URLs, decodes them into CGImages, and notifies
/// the caller with the decoded image and its pixel dimensions.
class ImageLoader {
    private let onImageLoaded: (_ imageId: String, _ image: CGImage, _ width: CGFloat, _ height: CGFloat) -> Void

    /// Cache of decoded images keyed by imageId.
    var cache: [String: CGImage] = [:]

    init(onImageLoaded: @escaping (_ imageId: String, _ image: CGImage, _ width: CGFloat, _ height: CGFloat) -> Void) {
        self.onImageLoaded = onImageLoaded
    }

    func load(imageId: String, url: String) {
        guard let imageUrl = URL(string: url) else { return }
        URLSession.shared.dataTask(with: imageUrl) { [weak self] data, _, error in
            guard let self = self,
                  let data = data, error == nil,
                  let source = CGImageSourceCreateWithData(data as CFData, nil),
                  let cgImage = CGImageSourceCreateImageAtIndex(source, 0, nil)
            else {
                if let error = error {
                    NSLog("[ImageLoader] Failed to load image %@: %@", imageId, error.localizedDescription)
                }
                return
            }
            let width = CGFloat(cgImage.width)
            let height = CGFloat(cgImage.height)
            self.cache[imageId] = cgImage
            self.onImageLoaded(imageId, cgImage, width, height)
        }.resume()
    }
}

File: native/android/app/src/main/java/io/johnsonlee/glyphis/shell/ImageLoader.kt
package io.johnsonlee.glyphis.shell

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import java.net.URL
import java.util.concurrent.Executors

/**
 * Loads images from URLs, decodes them into Bitmaps, and notifies
 * the caller with the decoded dimensions (in dp).
 *
 * Uses a fixed thread pool for concurrent loading with bounded parallelism.
 * Deduplicates requests — multiple loads for the same imageId are ignored
 * if a load is already in progress or the image is cached.
 */
class ImageLoader(
    private val density: Float,
    private val onImageLoaded: (imageId: String, bitmap: Bitmap, widthDp: Float, heightDp: Float) -> Unit,
) {
    val cache = mutableMapOf<String, Bitmap>()
    private val pending = mutableSetOf<String>()
    private val executor = Executors.newFixedThreadPool(4)

    fun load(imageId: String, url: String) {
        // Cache hit — deliver immediately
        val cached = cache[imageId]
        if (cached != null) {
            onImageLoaded(
                imageId, cached,
                cached.width.toFloat() / density,
                cached.height.toFloat() / density,
            )
            return
        }

        // Deduplicate — skip if already loading
        synchronized(pending) {
            if (!pending.add(imageId)) return
        }

        executor.execute {
            try {
                URL(url).openStream().use { stream ->
                    val bitmap = BitmapFactory.decodeStream(stream)
                    if (bitmap != null) {
                        cache[imageId] = bitmap
                        onImageLoaded(
                            imageId, bitmap,
                            bitmap.width.toFloat() / density,
                            bitmap.height.toFloat() / density,
                        )
                    } else {
                        Log.w(TAG, "Failed to decode image: $url")
                    }
                }
            } catch (e: Exception) {
                Log.w(TAG, "Failed to load image: $url", e)
            } finally {
                synchronized(pending) {
                    pending.remove(imageId)
                }
            }
        }
    }

    fun destroy() {
        executor.shutdownNow()
        cache.clear()
        synchronized(pending) { pending.clear() }
    }

    companion object {
        private const val TAG = "ImageLoader"
    }
}
